pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/rebalance/ReplicaGroupRebalanceSegmentStrategy.java

```diff
@@ -0,0 +1,573 @@
+                    + "Also, please consider to use BOOTSTRAP mode.");
+        }
+        if (!removeGroup) {
+          newReplicaGroupPartitionAssignment.setInstancesToReplicaGroup(partitionId, groupId, newReplicaGroup);
+        }
+      }
+
+      // Adding new replica groups if needed
+      int index = 0;
+      for (int newGroupId = oldNumReplicaGroup; newGroupId < targetNumReplicaGroup; newGroupId++) {
+        List<String> newReplicaGroup = new ArrayList<>();
+        while (newReplicaGroup.size() < targetNumInstancesPerPartition) {
+          newReplicaGroup.add(addedServers.get(index));
+          index++;
+        }
+        newReplicaGroupPartitionAssignment.setInstancesToReplicaGroup(partitionId, newGroupId, newReplicaGroup);
+      }
+    }
+
+    return newReplicaGroupPartitionAssignment;
+  }
+
+  /**
+   * Given the list of added/removed servers and the old and new replica group configurations, compute the
+   * type of update (e.g. replace server, add servers to each replica group, add replica groups..)
+   *
+   * @return the update type
+   */
+  private ReplicaGroupRebalanceType computeRebalanceType(int oldNumInstancesPerPartition, int oldNumReplicaGroup,
+      int targetNumReplicaGroup, int targetNumInstancesPerPartition, List<String> addedServers,
+      List<String> removedServers) {
+    boolean sameNumInstancesPerPartition = oldNumInstancesPerPartition == targetNumInstancesPerPartition;
+    boolean sameNumReplicaGroup = oldNumReplicaGroup == targetNumReplicaGroup;
+    boolean isAddedServersSizeZero = addedServers.size() == 0;
+    boolean isRemovedServersSizeZero = removedServers.size() == 0;
+
+    if (sameNumInstancesPerPartition && sameNumReplicaGroup && addedServers.size() == removedServers.size()) {
+      return ReplicaGroupRebalanceType.REPLACE;
+    } else if (sameNumInstancesPerPartition) {
+      if (oldNumReplicaGroup < targetNumReplicaGroup && !isAddedServersSizeZero && isRemovedServersSizeZero
+          && addedServers.size() % targetNumInstancesPerPartition == 0) {
+        return ReplicaGroupRebalanceType.ADD_REPLICA_GROUP;
+      } else if (oldNumReplicaGroup > targetNumReplicaGroup && isAddedServersSizeZero && !isRemovedServersSizeZero
+          && removedServers.size() % targetNumInstancesPerPartition == 0) {
+        return ReplicaGroupRebalanceType.REMOVE_REPLICA_GROUP;
+      }
+    } else if (sameNumReplicaGroup) {
+      if (oldNumInstancesPerPartition < targetNumInstancesPerPartition && isRemovedServersSizeZero
+          && !isAddedServersSizeZero && addedServers.size() % targetNumReplicaGroup == 0) {
+        return ReplicaGroupRebalanceType.ADD_SERVER;
+      } else if (oldNumInstancesPerPartition > targetNumInstancesPerPartition && !isRemovedServersSizeZero
+          && isAddedServersSizeZero && removedServers.size() % targetNumReplicaGroup == 0) {
+        return ReplicaGroupRebalanceType.REMOVE_SERVER;
+      }
+    }
+    return ReplicaGroupRebalanceType.UNSUPPORTED;
```

Comment:
You should at least log an INFO with the value of variables computed, and why the code has reached this conclusion.  It will help us debug when an operator claims that replica group rebalance does not work. Ideally, provide that error back to the admin as well, so that they know what they are doing wrong (e.g. trying to do multiple operations in one go). We can avoid one call that way :)

Reply:
added log